### ISSUE TITLE
'tesseract::Dawg::iterate_words': ambiguous call to overloaded function

### ISSUE THREAD

**Issue by z-pc** [tesseract-ocr/tesseract#3107]
### Current Behavior:
1. Download a latest sw version.
2. Run list cmd:
```
sw setup
sw build org.sw.demo.google.tesseract.tesseract-master
git clone https://github.com/tesseract-ocr/tesseract tesseract
cd tesseract
mkdir build && cd build
cmake .. -G "Visual Studio 14 2015" -DCMAKE_INSTALL_PREFIX=inst
cmake --build . --config Release
```

All success, except "dawg2wordlist" project.
I tried build again by VS2015 IDE. Still error:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error (active)		more than one instance of overloaded function "tesseract::Dawg::iterate_words" matches the argument list:	dawg2wordlist	g:\buildproject\tesseract\tesseract\src\training\dawg2wordlist.cpp	64	

Severity	Code	Description	Project	File	Line	Suppression State
Error	C2668	'tesseract::Dawg::iterate_words': ambiguous call to overloaded function	dawg2wordlist	G:\buildproject\tesseract\tesseract\src\training\dawg2wordlist.cpp	65	
```
This is a detail log on Output console:
[output.log](https://github.com/tesseract-ocr/tesseract/files/5290500/output.log)

------------------------

### Environment

* **Tesseract Version**: 5.0
* **Commit Number**: cec7563
* **Platform**: build by VS 2015, expected target on x64 and x86, Win10 x64
* **sw --version**:
```
sw.client.sw version 1.0.0
git revision 1fe6d9946df29af1295194e8b5712abb2967ceb3
assembled on
26.09.2020 22:13:22 UTC
27.09.2020 05:13:22 SE Asia Standard Time
```

**Comment by z-pc** [comment]
Sorry, I don't really understand how sw works yet.
If I run:
```
sw build org.sw.demo.google.tesseract.tesseract-master
```
Does this mean that sw helped me to build the library? And do I just link them to my project?
In other word if I want to the latest library, built directly from the source code, I don't need to run it?

**Comment by Moldoteck** [comment]
If you want the last version from git for tesseract, sw will not help you without additional tweaks. 
sw build org.sw.demo.google.tesseract.tesseract-master will download the source code and build it but the code will be from 2020-03-17 commit(i think it's last comit for 4.1 or 4.0 version).
You can inspect the source directory or build directory of sw by clicking on buttons here: https://software-network.org/org.sw.demo.google.tesseract.libtesseract/actions.

If you want to use tesseract in your code, you can use sw in cmake, for example, add this in main cmake file:
```

set(SW_BUILD_SHARED_LIBS 0)
set(LIBTESSERACT
    org.sw.demo.danbloomberg.leptonica
    org.sw.demo.google.tesseract.libtesseract-master
)
find_package(SW REQUIRED)
sw_add_package(${LIBTESSERACT})
sw_execute()

```

Later in the project, you can just do something like

`target_link_libraries(YourTarget PUBLIC ${LIBTESSERACT})`

**Comment by z-pc** [comment]
I understand what your mean, but I see this on CMakelist of tesseract git:
```
   sw_add_package(
        org.sw.demo.danbloomberg.leptonica
        org.sw.demo.libarchive.libarchive
    )
    if (BUILD_TRAINING_TOOLS)
        sw_add_package(
            org.sw.demo.gnome.pango.pangocairo
            org.sw.demo.unicode.icu.i18n
        )
    endif()
```
It seems can still build directly tesseract from git without related to the built-in tesseract by sw ?

**Comment by Moldoteck** [comment]
Yes, in this case tesseract will be build using it's cmake, sw and external dependencies(like leptonica, libarchieve).

But, if you run sw build org.sw.demo.google.tesseract.tesseract-master, sw will build itself the tesseract by using an internal build script(it is used instead of tesseract's cmake).

It's up to you what way to use. 
Your error is related to first way of building tesseract. (Maybe it is specific to version 5.0.0 alpha)

**Comment by z-pc** [comment]
Thanks for respond.
I tried delete C:\User\ABC\.sw dir, run uninstall sw, delete old .sw directory in build dir directory , and delete git tesseract dir.
Run again all without `sw build org.sw.demo.google.tesseract.tesseract-master`.
It's still error.

**Comment by z-pc** [comment]
I have no problem with using older versions. But the tricky thing for me is that my application was long ago created and developed using just VS2015, without CMakelist or sw before it. It is only built for use on Windows. I think using vcpkg would be more appropriate for my case.

**Comment by zdenop** [comment]
vcpkg will build static version of tesseract on Windows. And  AFAIK vcpkg  use cmake on background.
You can build tesseract by yourself without sw (but without training tools on windows).
Cmake  should be able to generate  solution for installed VS. Or have a look [at this tutorial](http://spell.linux.sk/building-tesseract-and-leptonica-with-cmake-and-clang-on-windows) - just replace clang with VS...

**Comment by z-pc** [comment]
Thanks for recommend @zdenop.
I solved this problems. I did:

Change source code
`
  dawg->iterate_words(unicharset,
				     std::bind(&WordOutputter::output_word, &outputter, _1));
`
to 
`
  dawg->iterate_words(unicharset,
					(std::function<void(const char*)>)std::bind(&WordOutputter::output_word, &outputter, _1));
`

I don't know why my VS2015 not link to 
`
void Dawg::iterate_words(const UNICHARSET& unicharset,
                         std::function<void(const char*)> cb)
`

Then build solution again, tt's worked. Next I run tesseract.exe with cmd:
`
tesseract.exe E:\input.png output --oem 1 -l eng
`

I receive this error on tesseract x86 (x64 and VS2019 is no problem)
`
tesseract the application was unable to start correctly "0xc00007b"
`

So I installed a CMake latest that is "cmake-3.19.0-rc1-win64-x64", before is cmake-3.14.0-rc2-win32-x86. Configure and generate solution with **CMake GUI** (it's seem CMake 3.19 has a few small changes to the default arguments) with previously changed source code.
I don't know why, but it's worked, not have a "0xc00007b" error.
